src/System.Text.Encoding.CodePages/tests/EncoderFallbackBufferHelper.cs

```diff
@@ -0,0 +1,88 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Xunit;
+
+namespace System.Text.Tests
+{
+    public class TestEncoderFallbackBuffer : EncoderFallbackBuffer
+    {
+        private int _fallbackCount = 0;
+
+        public TestEncoderFallbackBuffer() { }
+
+        public override bool Fallback(char charUnknownHigh, char charUnknownLow, int index)
+        {
+            _fallbackCount++;
+            return true;
+        }
+
+        public override bool Fallback(char charUnknown, int index)
+        {
+            _fallbackCount++;
+            return true;
+        }
+
+        public override bool MovePrevious()
+        {
+            if (_fallbackCount > 0)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        public override int Remaining { get { return _fallbackCount; } }
+
+        public override char GetNextChar()
+        {
+            if (_fallbackCount > 0)
+            {
+                _fallbackCount--;
+                return '?';
+            }
+            return '\0';
+        }
+    }
+
+    public class TestEncoderFallback : EncoderFallback
+    {
+        public TestEncoderFallback() { }
+
+        public override EncoderFallbackBuffer CreateFallbackBuffer()
+        {
+            return new TestEncoderFallbackBuffer();
+        }
+
+        public override int MaxCharCount { get { return 2; } }
+    }
+
+    public class EncoderFallbackBufferHelperTest : RemoteExecutorTestBase
+    {
+        [Fact]
+        public void TestEncoderFallbackBufferHelper()
+        {
+            RemoteInvoke(() =>
+            {
+                // Add the code page provider.
+                Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
+
+                // Use windows-1252 with a custom fallback that utilizes the EncoderFallbackBufferHelper.
+                Encoding encoding = Encoding.GetEncoding(1252, new TestEncoderFallback(), DecoderFallback.ReplacementFallback);
+
+                // Roundtrip unsupported character.
+                byte[] bytes = encoding.GetBytes("\u0661");
+                char[] chars = encoding.GetChars(bytes);
+                Assert.Equal("?", new string(chars));
+
+                // Roundtrip a high surrogate character.
+                bytes = encoding.GetBytes("\uD800");
+                chars = encoding.GetChars(bytes);
+                Assert.Equal("?", new string(chars));
```

Comment:
It's great that this causes the various code paths to be executed, but that doesn't necessarily mean the right things are happening.  For example, can/should this validate that the expected inputs are being passed into the various methods on the EncoderFallbackBuffer?